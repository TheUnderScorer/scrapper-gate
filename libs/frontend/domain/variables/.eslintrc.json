{
  "extends": ["plugin:@nrwl/nx/react", "../../../../.eslintrc.json"],
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "parserOptions": {
        "project": [
          "libs/frontend/domain/variables/tsconfig.*?.json",
          "libs/frontend/domain/variables/.storybook/tsconfig.json"
        ]
      },
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    }
  ]
}
