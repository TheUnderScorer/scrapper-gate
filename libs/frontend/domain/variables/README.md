# frontend-domain-variables

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test frontend-domain-variables` to execute the unit tests via [Jest](https://jestjs.io).
