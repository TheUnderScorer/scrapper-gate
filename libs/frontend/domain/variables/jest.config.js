module.exports = {
  displayName: 'frontend-domain-variables',
  preset: '../../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../../coverage/libs/frontend/domain/variables',
  setupFilesAfterEnv: [`${process.cwd()}/tests/setupFrontendTests.ts`],
};
