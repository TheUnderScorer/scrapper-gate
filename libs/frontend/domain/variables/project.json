{
  "root": "libs/frontend/domain/variables",
  "sourceRoot": "libs/frontend/domain/variables/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/frontend/domain/variables/**/*.{ts,tsx,js,jsx}"
        ]
      },
      "outputs": ["{options.outputFile}"]
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/frontend/domain/variables"],
      "options": {
        "jestConfig": "libs/frontend/domain/variables/jest.config.js",
        "passWithNoTests": true
      }
    },
    "storybook": {
      "executor": "@nrwl/storybook:storybook",
      "options": {
        "uiFramework": "@storybook/react",
        "port": 4400,
        "config": {
          "configFolder": "libs/frontend/domain/variables/.storybook"
        }
      },
      "configurations": {
        "ci": {
          "quiet": true
        }
      }
    },
    "build-storybook": {
      "executor": "@nrwl/storybook:build",
      "outputs": ["{options.outputPath}"],
      "options": {
        "uiFramework": "@storybook/react",
        "outputPath": "dist/storybook/frontend-domain-variables",
        "config": {
          "configFolder": "libs/frontend/domain/variables/.storybook"
        }
      },
      "configurations": {
        "ci": {
          "quiet": true
        }
      }
    }
  },
  "tags": []
}
