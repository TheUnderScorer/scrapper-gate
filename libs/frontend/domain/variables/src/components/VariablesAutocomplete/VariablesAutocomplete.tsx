import {
  Autocomplete,
  AutocompleteProps,
  AutocompleteRenderInputParams,
  createFilterOptions,
  FilterOptionsState,
} from '@mui/material';
import { FieldProps, useFieldHasError } from '@scrapper-gate/frontend/form';
import { Highlight } from '@scrapper-gate/frontend/ui';
import { Variable } from '@scrapper-gate/shared/schema';
import React, { useCallback } from 'react';
import { FieldInputProps, FieldMetaState, useField } from 'react-final-form';
import { useVariablesContextSelector } from '../../providers/VariablesProvider';

export interface VariablesAutocompleteChildrenBag {
  autoCompleteParams: AutocompleteRenderInputParams;
  meta: FieldMetaState<string>;
  input: FieldInputProps<string>;
  hasError?: boolean;
}

export interface VariablesAutocompleteProps
  extends Omit<
    AutocompleteProps<Variable, boolean, boolean, boolean>,
    'options' | 'renderInput'
  > {
  children: (params: VariablesAutocompleteChildrenBag) => JSX.Element;
  fieldProps: FieldProps<string>;
  name: string;
}

export const VariablesAutocomplete = ({
  name,
  children,
  fieldProps,
  ...rest
}: VariablesAutocompleteProps) => {
  const { meta, input } = useField(name, fieldProps);

  const hasError = useFieldHasError({
    meta: meta,
    showErrorOnlyOnTouched: fieldProps?.showErrorOnlyOnTouched,
  });

  const variables = useVariablesContextSelector((ctx) => ctx.variables);

  const filterOptions = useCallback(
    (options: Variable[], state: FilterOptionsState<Variable>) => {
      if (!state.inputValue.startsWith('{{')) {
        return [];
      }

      return createFilterOptions<Variable>({
        matchFrom: 'any',
        stringify: (option) => option.key ?? option.id,
      })(options, state);
    },
    []
  );

  return (
    <Autocomplete<Variable, boolean, boolean, boolean>
      {...rest}
      {...input}
      onChange={(event, newValue) => input.onChange(newValue)}
      id={input.name}
      freeSolo
      disableClearable
      renderInput={(params) =>
        children({
          autoCompleteParams: params,
          meta,
          input,
          hasError,
        })
      }
      getOptionLabel={(option) => {
        return option?.key ?? '';
      }}
      filterOptions={filterOptions}
      options={variables}
      renderOption={(props, variable, state) => {
        return (
          <Highlight text={variable.key ?? ''} value={state.inputValue ?? ''} />
        );
      }}
    />
  );
};
