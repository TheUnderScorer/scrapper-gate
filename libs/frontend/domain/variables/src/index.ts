export * from './lib/frontend-domain-variables';
