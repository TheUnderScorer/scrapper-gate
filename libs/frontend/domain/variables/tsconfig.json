{
  "extends": "../../../../tsconfig.base.json",
  "compilerOptions": {
    "jsx": "react-jsx",
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true
  },
  "files": ["../../block-editor/src/typings/slate/index.d.ts"],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    },
    {
      "path": "./.storybook/tsconfig.json"
    }
  ]
}
