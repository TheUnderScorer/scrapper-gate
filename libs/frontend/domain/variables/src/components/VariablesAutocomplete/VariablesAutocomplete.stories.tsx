import { Box } from '@mui/material';
import { FormVariablesProvider } from '../../providers/Form/FormVariablesProvider';
import { createVariable } from '@scrapper-gate/shared/domain/variables';
import { Variable, VariableScope } from '@scrapper-gate/shared/schema';
import { Form } from 'react-final-form';
import { VariablesAutocomplete } from './VariablesAutocomplete';

export default {
  title: 'Variables Autocomplete',
};

const variables: Variable[] = [
  createVariable({
    key: 'Myvariable',
    value: 'Variable test',
    defaultValue: 'Test',
    scope: VariableScope.Global,
  }),
  createVariable({
    key: 'Testvar',
    defaultValue: 'Def',
    scope: VariableScope.Global,
  }),
  createVariable({
    key: 'Longvariablename',
    defaultValue: 'Default',
    scope: VariableScope.Scrapper,
  }),
];

const options = ['Test', 'Another'];

export const Component = () => {
  return (
    <Form
      initialValues={{
        variables,
      }}
      onSubmit={console.log}
      render={() => (
        <Box width="600px">
          <FormVariablesProvider name="variables">
            <VariablesAutocomplete name="test" options={options} />
          </FormVariablesProvider>
        </Box>
      )}
    />
  );
};
