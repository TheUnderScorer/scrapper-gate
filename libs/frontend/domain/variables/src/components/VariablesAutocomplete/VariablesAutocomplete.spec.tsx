/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { createVariable } from '@scrapper-gate/shared/domain/variables';
import { VariableScope, VariableType } from '@scrapper-gate/shared/schema';
import { act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Transforms } from 'slate';
import {
  RenderBlockEditorComponent,
  renderVariablesBlockEditor,
} from '../../../../../../../tests/ui/blockEditor/renderBlockEditor';
import { VariablesAutocomplete } from './VariablesAutocomplete';

const variables = [
  createVariable({
    key: 'Test',
    defaultValue: '123',
    type: VariableType.Text,
    scope: VariableScope.Global,
  }),
];

const options = ['Test', 'Cool', 'Another'];

const Component: RenderBlockEditorComponent = (props) => (
  <VariablesAutocomplete {...props} options={options} />
);

describe('<VariablesAutocomplete />', () => {
  it('should render without crashing', async () => {
    const { cmp } = await renderVariablesBlockEditor({
      variables,
      Component,
    });

    expect(cmp).toMatchSnapshot();
  });

  it('should show suggestions', async () => {
    const { cmp } = await renderVariablesBlockEditor({
      variables,
      Component,
    });

    expect(
      cmp.container.querySelector('.variables-autocomplete-suggestions')
    ).toBeInTheDocument();

    expect(
      cmp.container.querySelectorAll('.variables-autocomplete-suggestion')
    ).toHaveLength(options.length);
  });

  it('should show filtered suggestions', async () => {
    const { cmp } = await renderVariablesBlockEditor({
      variables,
      Component,
      initialValue: 'Tes',
    });

    const suggestions = cmp.container.querySelectorAll(
      '.variables-autocomplete-suggestion'
    );

    expect(suggestions).toHaveLength(1);
    expect(suggestions[0]).toHaveTextContent(/Test/);
  });

  it('should set text field value to clicked suggestion', async () => {
    const { cmp, editor } = await renderVariablesBlockEditor({
      variables,
      Component,
      initialValue: 'Tes',
    });

    const suggestion = cmp.container.querySelector(
      '.variables-autocomplete-suggestion'
    )!;

    act(() => {
      Transforms.select(editor, {
        anchor: {
          path: [0, 0],
          offset: 0,
        },
        focus: {
          path: [0, 0],
          offset: 2,
        },
      });
    });

    act(() => {
      userEvent.click(suggestion);
    });

    expect(
      cmp.container.querySelector('[contenteditable=true]')
    ).toHaveTextContent(/Test/);
  });
});
