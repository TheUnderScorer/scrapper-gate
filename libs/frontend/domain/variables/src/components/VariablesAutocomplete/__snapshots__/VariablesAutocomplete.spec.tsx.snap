// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VariablesAutocomplete /> should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    >
      <div
        class="MuiBox-root css-8atqhb"
      >
        <div
          class="MuiFormControl-root MuiTextField-root block-editor-text-field css-9e6gxp-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-11zt3ho-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiBox-root css-1yyq5vx"
            >
              <div
                aria-invalid="false"
                aria-multiline="false"
                autocapitalize="false"
                autocorrect="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                contenteditable="true"
                data-gramm="false"
                data-slate-editor="true"
                data-slate-node="value"
                name="variable"
                role="textbox"
                spellcheck="false"
                style="position: relative; outline: none; white-space: pre-wrap; word-wrap: break-word;"
                type="text"
              >
                <div
                  data-slate-node="element"
                  style="position: relative;"
                >
                  <span
                    data-slate-node="text"
                  >
                    <span
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-length="0"
                        data-slate-zero-width="n"
                      >
                        ﻿
                        <br />
                      </span>
                    </span>
                  </span>
                </div>
              </div>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-nnbavb"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiModal-root MuiPopover-root variables-autocomplete-suggestions css-18l9qlm-MuiModal-root-MuiPopover-root-MuiMenu-root"
          role="presentation"
          style="pointer-events: all;"
        >
          <div
            aria-hidden="true"
            class="MuiBackdrop-root MuiBackdrop-invisible css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
            style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          />
          <div
            data-test="sentinelStart"
            tabindex="0"
          />
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiMenu-paper MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-1poimk-MuiPaper-root-MuiMenu-paper-MuiPaper-root-MuiPopover-paper"
            style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
            tabindex="-1"
          >
            <ul
              class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
              role="menu"
              tabindex="-1"
            >
              <li
                class="MuiMenuItem-root MuiMenuItem-gutters Mui-selected MuiButtonBase-root variables-autocomplete-suggestion selected css-soz1h5-MuiButtonBase-root-MuiMenuItem-root"
                role="menuitem"
                tabindex="0"
              >
                <span>
                  Test
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </li>
              <li
                class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root variables-autocomplete-suggestion css-soz1h5-MuiButtonBase-root-MuiMenuItem-root"
                role="menuitem"
                tabindex="-1"
              >
                <span>
                  Cool
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </li>
              <li
                class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root variables-autocomplete-suggestion css-soz1h5-MuiButtonBase-root-MuiMenuItem-root"
                role="menuitem"
                tabindex="-1"
              >
                <span>
                  Another
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </li>
            </ul>
          </div>
          <div
            data-test="sentinelEnd"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div
      class="MuiBox-root css-8atqhb"
    >
      <div
        class="MuiFormControl-root MuiTextField-root block-editor-text-field css-9e6gxp-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-11zt3ho-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            class="MuiBox-root css-1yyq5vx"
          >
            <div
              aria-invalid="false"
              aria-multiline="false"
              autocapitalize="false"
              autocorrect="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              contenteditable="true"
              data-gramm="false"
              data-slate-editor="true"
              data-slate-node="value"
              name="variable"
              role="textbox"
              spellcheck="false"
              style="position: relative; outline: none; white-space: pre-wrap; word-wrap: break-word;"
              type="text"
            >
              <div
                data-slate-node="element"
                style="position: relative;"
              >
                <span
                  data-slate-node="text"
                >
                  <span
                    data-slate-leaf="true"
                  >
                    <span
                      data-slate-length="0"
                      data-slate-zero-width="n"
                    >
                      ﻿
                      <br />
                    </span>
                  </span>
                </span>
              </div>
            </div>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-nnbavb"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiModal-root MuiPopover-root variables-autocomplete-suggestions css-18l9qlm-MuiModal-root-MuiPopover-root-MuiMenu-root"
        role="presentation"
        style="pointer-events: all;"
      >
        <div
          aria-hidden="true"
          class="MuiBackdrop-root MuiBackdrop-invisible css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
          style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        />
        <div
          data-test="sentinelStart"
          tabindex="0"
        />
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiMenu-paper MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-1poimk-MuiPaper-root-MuiMenu-paper-MuiPaper-root-MuiPopover-paper"
          style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
          tabindex="-1"
        >
          <ul
            class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
            role="menu"
            tabindex="-1"
          >
            <li
              class="MuiMenuItem-root MuiMenuItem-gutters Mui-selected MuiButtonBase-root variables-autocomplete-suggestion selected css-soz1h5-MuiButtonBase-root-MuiMenuItem-root"
              role="menuitem"
              tabindex="0"
            >
              <span>
                Test
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
            <li
              class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root variables-autocomplete-suggestion css-soz1h5-MuiButtonBase-root-MuiMenuItem-root"
              role="menuitem"
              tabindex="-1"
            >
              <span>
                Cool
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
            <li
              class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root variables-autocomplete-suggestion css-soz1h5-MuiButtonBase-root-MuiMenuItem-root"
              role="menuitem"
              tabindex="-1"
            >
              <span>
                Another
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
          </ul>
        </div>
        <div
          data-test="sentinelEnd"
          tabindex="0"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
